relayer/src/worker/packet.rs

```diff
@@ -109,12 +109,13 @@ impl PacketWorker {
                     height,
                     new_block: _,
                 } => {
-                    // Schedule the clearing of pending packets
+                    // Schedule the clearing of pending packets at start and
                     // at predefined block intervals.
-                    if self.clear_packets_interval != 0
-                        && height.revision_height % self.clear_packets_interval == 0
+                    if link.a_to_b.clear_packets()
```

Comment:
Two remarks for this change:

1. It can trigger `link.a_to_b.do_clear_packets` repeatedly until the `clear_packets` flag is set to false. This flag is set to false only when the worker receives a command with a batch of events (that is the first match arm above, on [line 103](https://github.com/informalsystems/ibc-rs/pull/1201/files#diff-d7d106630bd632b82613cdf6ec54b5623af0041717ff82a92b7f8579da2629cfR103)).

2. If I understand correctly, with this change there may be two separate invocations to `link.a_to_b.do_clear_packets`:

    - one invocation upon NewBlock (this call site), which would clear packets for that `NewBlock`'s height
    - another invocation with the call site within `update_schedule(batch)`, which would clear packets for the height of that `batch`, see the snippet below:

https://github.com/informalsystems/ibc-rs/blob/872a5030f9bcc46cfbbb71803554731b3a12ca16/relayer/src/link/relay_path.rs#L299-L306

If the height of `NewBlock` event and the `batch` height are the same, then we'll clear packets twice for the same height. Worst case, this will schedule twice the execution of the same operational data.

BTW I tried to reproduce point (2) but was not able. I was only able to reproduce problem (1) above.

Reply:
I attempted an alternative fix
 628425b

Feel free to revert if this approach is not appropriate.

Reply:
> If the height of NewBlock event and the batch height are the same, then we'll clear packets twice for the same height. Worst case, this will schedule twice the execution of the same operational data.

Ah yes, good point! This is the case in v0.5.0 and older also. We used to get the IBC event before the New Block event so we had to clear on IBC event also. With the changes in the scheduler and monitor we now always get the NewBlock before the IBC event. I think we can clear packets on NewBlock only. I will try some changes along these lines so we simplify the code quite a bit.

>  I was only able to reproduce problem (1) above.
How?

Reply:
> How?

Steps to reproduce (1)

1. dev-env with two chains, and create a channel between the two;
2. `git co 872a503` and `start` hermes. The logs show

>Jul 15 09:24:16.570  INFO [ibc-1:transfer/channel-0 -> ibc-0] clearing pending packets from events before height 1-312
Jul 15 09:24:16.581  INFO [ibc-1:transfer/channel-0 -> ibc-0] finished scheduling the clearing of pending packets
Jul 15 09:24:16.886  INFO [ibc-0:transfer/channel-0 -> ibc-1] clearing pending packets from events before height 0-324
Jul 15 09:24:16.902  INFO [ibc-0:transfer/channel-0 -> ibc-1] finished scheduling the clearing of pending packets
Jul 15 09:24:17.723  INFO [ibc-1:transfer/channel-0 -> ibc-0] clearing pending packets from events before height 1-313
Jul 15 09:24:17.733  INFO [ibc-1:transfer/channel-0 -> ibc-0] finished scheduling the clearing of pending packets
Jul 15 09:24:18.091  INFO [ibc-0:transfer/channel-0 -> ibc-1] clearing pending packets from events before height 0-325
Jul 15 09:24:18.101  INFO [ibc-0:transfer/channel-0 -> ibc-1] finished scheduling the clearing of pending packets
Jul 15 09:24:18.872  INFO [ibc-1:transfer/channel-0 -> ibc-0] clearing pending packets from events before height 1-314
Jul 15 09:24:18.880  INFO [ibc-1:transfer/channel-0 -> ibc-0] finished scheduling the clearing of pending packets
Jul 15 09:24:19.190  INFO [ibc-0:transfer/channel-0 -> ibc-1] clearing pending packets from events before height 0-326

The problem was: the flag `link.a_to_b.clear_packets` is never set to `false`, so upon every `NewBlock` event, the method `do_clear_packets` is invoked.